svm/src/account_loader.rs

```diff
@@ -95,6 +96,264 @@ pub struct FeesOnlyTransaction {
     pub fee_details: FeeDetails,
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct AccountLoader<'a, CB: TransactionProcessingCallback> {
+    pub program_cache: ProgramCacheForTxBatch,
+    account_overrides: Option<&'a AccountOverrides>,
+    account_cache: HashMap<Pubkey, AccountCacheItem>,
+    callbacks: &'a CB,
+}
+impl<'a, CB: TransactionProcessingCallback> AccountLoader<'a, CB> {
+    pub fn new(
+        callbacks: &'a CB,
+        account_overrides: Option<&'a AccountOverrides>,
+        program_cache: ProgramCacheForTxBatch,
+    ) -> AccountLoader<'a, CB> {
+        Self {
+            program_cache,
+            account_overrides,
+            account_cache: HashMap::new(),
+            callbacks,
+        }
+    }
+
+    pub fn load_account(
+        &mut self,
+        account_key: &Pubkey,
+        usage_pattern: AccountUsagePattern,
+    ) -> Option<LoadedTransactionAccount> {
+        let is_writable = usage_pattern.is_writable();
+        let is_invisible = usage_pattern.is_invisible();
+
+        if let Some(cache_item) = self.account_cache.get_mut(account_key) {
+            if !cache_item.inspected_as_writable {
+                // Inspecting a read-only account is harmless. Inspecting a writable
+                // account must be done before rent collection, and must never be done
+                // again in-batch, because inspection is intended to preserve the
+                // state prior to any writes.
+                self.callbacks.inspect_account(
+                    account_key,
+                    AccountState::Alive(&cache_item.account),
+                    is_writable,
+                );
+                cache_item.inspected_as_writable = is_writable;
+            }
+
+            // We must compare the full account, not just check if lamports are 0.
+            // Otherwise we might erroneously hide rent-delinquent read-only accounts.
+            if cache_item.account == AccountSharedData::default() {
```

Comment:
`AccountSharedData::default()` will allocate the `Arc`; we should have some `is_default_account` fn that makes the comparisons without allocating

Maybe outside the scope of this PR.

Reply:
for now im changing this to use `accounts_equal()`. the code works fine with `==`, but when i was researching `Arc` i learned that theyre compared using pointer equality, and the reason _why_ this code works is because rustc optimizes empty vecs to point to the same location. i assume this would never change but its better to be sure

i could pr to add `is_default()` to `AccountSharedData` separately from this. neither pr would block the other, if this pr goes in first we can just change it after

Reply:
[6d9041f](https://github.com/anza-xyz/agave/pull/3146/commits/6d9041f0dfbcd32457284e3378cee6bc91c23e90)